pyro/ops/gaussian.py

```diff
@@ -34,17 +35,72 @@ def __init__(self, log_normalizer, info_vec, precision):
     def dim(self):
         return self.info_vec.size(-1)
 
-    @property
+    @lazy_property
     def batch_shape(self):
         return broadcast_shape(self.log_normalizer.shape,
                                self.info_vec.shape[:-1],
                                self.precision.shape[:-2])
 
+    def expand(self, batch_shape):
+        n = self.info_vec.size(-1)
+        log_normalizer = self.log_normalizer.expand(batch_shape)
+        info_vec = self.info_vec.expand(batch_shape + (n,))
+        precision = self.precision.expand(batch_shape + (n, n))
+        return Gaussian(log_normalizer, info_vec, precision)
+
+    def reshape(self, batch_shape):
+        n = self.info_vec.size(-1)
+        log_normalizer = self.log_normalizer.reshape(batch_shape)
+        info_vec = self.info_vec.reshape(batch_shape + (n,))
+        precision = self.precision.reshape(batch_shape + (n, n))
+        return Gaussian(log_normalizer, info_vec, precision)
+
+    def __getitem__(self, index):
+        """
+        Index into the batch_shape of a Gaussian.
+        """
+        assert isinstance(index, tuple)
+        log_normalizer = self.log_normalizer[index]
+        info_vec = self.info_vec[index + (slice(None),)]
+        precision = self.precision[index + (slice(None), slice(None))]
+        return Gaussian(log_normalizer, info_vec, precision)
+
+    @staticmethod
+    def cat(parts, dim=0):
+        """
+        Concatenate a list of Gaussians along a given batch dimension.
+        """
+        if dim < 0:
+            dim += len(parts[0].batch_shape)
+        args = [torch.cat([getattr(g, attr) for g in parts], dim=dim)
+                for attr in ["log_normalizer", "info_vec", "precision"]]
+        return Gaussian(*args)
+
+    def pad(self, left=0, right=0):
+        """
+        Pad along event dimension.
+        """
+        lr = (left, right)
+        log_normalizer = self.log_normalizer
+        info_vec = pad(self.info_vec, lr)
+        precision = pad(self.precision, lr + lr)
+        return Gaussian(log_normalizer, info_vec, precision)
+
+    def __add__(self, other):
+        """
+        Adds two Gaussians in log-density space.
+        """
+        assert self.dim() == other.dim()
+        return Gaussian(self.log_normalizer + other.log_normalizer,
+                        self.info_vec + other.info_vec,
+                        self.precision + other.precision)
+
     def log_density(self, value):
         """
-        Evaluate the log density of this Gaussian at a point value.
+        Evaluate the log density of this Gaussian at a point value::
```

Comment:
@fehiepsi Can you take a look at this change when you get a chance? I believe I have added a missing term, but I haven't been able to get all tests passing.

Reply:
I think that the additional term which you added does not depend on `value`, so would be absorbed in `log_normalizer` term.